File: AOC_2023/Day 15/15.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type lensSlot struct {
	focalLength int
	slot int
}

func initialize(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var hashes []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		for _, hash := range strings.Split(text, ",") {
			hashes = append(hashes, hash)
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return hashes, nil
}

func hashAlgo(input string) int {
	result := 0
	for _, char := range input {
		result += int(char)
		result *= 17
		result %= 256
	}
	return result
}

func part1(input []string) {
	result := 0
	for _, hash := range input {
		result += hashAlgo(hash)
	}
	fmt.Println("Part 1:", result)
}

func part2(input []string) {
	result := 0
	boxes := make([]map[string]*lensSlot, 256)
	for i := range boxes {
		boxes[i] = make(map[string]*lensSlot)
	}
	for _, hash := range input {
		if strings.Contains(hash, "=") {
			label := hash[:strings.Index(hash, "=")]
			boxNum := hashAlgo(label)
			focalLength := int(hash[len(hash)-1] - '0')
			if _, ok := boxes[boxNum][label]; ok {
				boxes[boxNum][label] = &lensSlot{focalLength, boxes[boxNum][label].slot}
			} else {
				boxes[boxNum][label] = &lensSlot{focalLength, len(boxes[boxNum])+1}
			}
		} else{
			label := hash[:strings.Index(hash, "-")]
			boxNum := hashAlgo(label)
			if _, ok := boxes[boxNum][label]; ok {
				pos := boxes[boxNum][label].slot
				delete(boxes[boxNum], label)
				for _, lens := range boxes[boxNum] {
					if lens.slot > pos {
						lens.slot -= 1
					}
				}
			}
		}
	}
	for i, box := range boxes {
		for _, lens := range box {
			result += (lens.focalLength * lens.slot*(i+1))
		}
	}
	fmt.Println("Part 2:", result)
}

func main(){
	filename := "input.txt"
	hashes, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	part1(hashes)
	part2(hashes)
}

File: AOC_2023/Day 15/15.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

class Solution
{
    public:
    vector<string> hashes;
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            stringstream ss(line);
            string token;
            while (getline(ss, token, ','))
            {
                hashes.push_back(token);
            }
        }
    }

    int hashAlgorithm(string hash){
        int result = 0;
        for (int i = 0; i < hash.length(); i++)
        {
            result += hash[i];
            result *= 17;
            result %= 256;
        }
        return result;
    }

    void part1(){
        int result = 0;
        for (int i = 0; i < hashes.size(); i++)
        {
            result += hashAlgorithm(hashes[i]);
        }
        cout<<"Part 1: "<<result<<"\n";
    }
    void part2(){
        int result = 0;
        vector<map<string,pair<int,int>>> boxes(256);
        for(auto hash : hashes){
            if(hash.find("=") != string::npos){
                int index = hash.find("=");
                string label = hash.substr(0,index);
                int boxNum = hashAlgorithm(label);
                int lens = stoi(hash.substr(index+1));
                if(boxes[boxNum].find(label) != boxes[boxNum].end()){
                    boxes[boxNum][label].first = lens;
                }
                else{
                    boxes[boxNum][label] = make_pair(lens, boxes[boxNum].size());
                }
            }
            else if(hash.find("-") != string::npos){
                int index = hash.find("-");
                string label = hash.substr(0,index);
                int boxNum = hashAlgorithm(label);
                if(boxes[boxNum].find(label) != boxes[boxNum].end()){
                    int pos = boxes[boxNum][label].second;
                    boxes[boxNum].erase(label);
                    for(auto it = boxes[boxNum].begin(); it != boxes[boxNum].end(); it++){
                        if(it->second.second > pos){
                            it->second.second--;
                        }
                    }
                }
            }
        }
        for(int i = 0; i < boxes.size(); i++){
            for(auto it = boxes[i].begin(); it != boxes[i].end(); it++){
                result += (it->second.second+1)*(i+1)*(it->second.first);
            }
        }
        cout<<"Part 2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}